csrc/multidevice/communication.cpp

```diff
@@ -111,209 +98,225 @@ Broadcast::Broadcast(CommParams params) : Communication(params, "broadcast") {}
 
 c10::intrusive_ptr<c10d::Work> Broadcast::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
 
   if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.src_bufs, 1);
-    if (params_.dst_bufs.size() == 1) {
-      doLocalCopy(params_.dst_bufs.at(0), params_.src_bufs.at(0));
+    if (params_.is_root_in_mesh) {
+      // Do a local copy and the subsequent broadcast will be in place.
+      doLocalCopy(output_tensor, input_tensor);
     } else {
-      assertBufferCount(params_.dst_bufs, 0);
+      // `output_tensor` isn't allocated for this device.
+      output_tensor = input_tensor;
     }
-  } else {
-    assertBufferCount(params_.src_bufs, 0);
-    assertBufferCount(params_.dst_bufs, 1);
   }
 
   if (params_.team.size() == 1) {
     return nullptr;
   }
 
+  std::vector<at::Tensor> tensors({output_tensor});
   return comm.getBackendForTeam(params_.team, backend)
-      ->broadcast(
-          comm.deviceId() == params_.root ? params_.src_bufs : params_.dst_bufs,
-          {.rootRank = root_relative_index_});
+      ->broadcast(tensors, {.rootRank = root_relative_index_});
 }
 
-Gather::Gather(CommParams params) : Communication(params, "gather") {
-  assertBufferCount(params_.src_bufs, 1);
-}
+Gather::Gather(CommParams params) : Communication(params, "gather") {}
 
 c10::intrusive_ptr<c10d::Work> Gather::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
-  // This is used to change the representation of the buffers to match c10d
-  // ProcessGroup API
-  std::vector<std::vector<at::Tensor>> buf_list;
-  if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.dst_bufs, params_.team.size());
-    buf_list = {std::move(params_.dst_bufs)};
-  } else {
-    assertBufferCount(params_.dst_bufs, 0);
+
+  if (comm.deviceId() == params_.root && !params_.is_root_in_mesh) {
+    input_tensor = at::empty_like(output_tensor.slice(0, 0, 1));
   }
-  auto work =
-      comm.getBackendForTeam(params_.team, backend)
-          ->gather(
-              buf_list, params_.src_bufs, {.rootRank = root_relative_index_});
+  std::vector<at::Tensor> input_tensors({input_tensor});
+
+  std::vector<std::vector<at::Tensor>> output_tensors;
   if (comm.deviceId() == params_.root) {
-    params_.dst_bufs = std::move(buf_list.back());
+    output_tensors.resize(1);
+    int64_t j = 0;
+    for (auto i : c10::irange(params_.team.size())) {
```

Comment:
I think you are right, and I'm surprised https://github.com/NVIDIA/Fuser/blob/4e8356915fc6ffa823cf97fed4249cc00f317fdf/tests/cpp/test_multidevice_pipeline.cpp#L218 doesn't fail. 

That being said, this appears to be a pre-existing issue according to https://github.com/NVIDIA/Fuser/blob/4e8356915fc6ffa823cf97fed4249cc00f317fdf/csrc/multidevice/lower_communication.cpp#L121. Am I right?

Reply:
I think the current code works correctly. The iteration is over the device mesh index so even if the mesh is {1,0} it will be ok! Perhaps just renaming i to mesh_idx will clear things up?